File: telegram_bot/services/product_service.py
from lib.mongodb import get_products_collection, get_categories_collection
from typing import List, Dict, Optional
from bson import ObjectId

class ProductService:
    @staticmethod
    async def get_all_products(category: Optional[str] = None, limit: int = 10, skip: int = 0) -> Dict:
        """Get all products with optional filtering"""
        collection = get_products_collection()

        query = {"isActive": True}
        if category:
            query["category"] = category

        total = await collection.count_documents(query)
        products = await collection.find(query).skip(skip).limit(limit).to_list(length=limit)

        return {
            "products": products,
            "total": total,
            "has_more": (skip + limit) < total
        }

    @staticmethod
    async def get_product_by_id(product_id: str) -> Optional[Dict]:
        """Get product by ID"""
        collection = get_products_collection()
        try:
            product = await collection.find_one({
                "_id": ObjectId(product_id),
                "isActive": True
            })
            return product
        except:
            return None

    @staticmethod
    async def search_products(query: str, limit: int = 10) -> List[Dict]:
        """Search products by name or description"""
        collection = get_products_collection()

        search_query = {
            "isActive": True,
            "$or": [
                {"name": {"$regex": query, "$options": "i"}},
                {"description": {"$regex": query, "$options": "i"}},
                {"tags": {"$in": [query]}}
            ]
        }

        products = await collection.find(search_query).limit(limit).to_list(length=limit)
        return products

    @staticmethod
    async def get_categories() -> List[Dict]:
        """Get all active categories"""
        collection = get_categories_collection()
        categories = await collection.find({"isActive": True}).sort("sortOrder", 1).to_list(length=100)
        return categories

    @staticmethod
    async def get_featured_products(limit: int = 5) -> List[Dict]:
        """Get featured products (high rating)"""
        collection = get_products_collection()

        products = await collection.find({
            "isActive": True,
            "rating": {"$gte": 4.5}
        }).sort("rating", -1).limit(limit).to_list(length=limit)

        return products

File: lib/services/productService.ts
import { ObjectId } from 'mongodb'
import { connectToDatabase } from '../mongodb'
import { Product, ProductView } from '../../types/database'

export class ProductService {
  static async getAllProducts(options: {
    category?: string
    limit?: number
    skip?: number
    sortBy?: 'price' | 'rating' | 'newest' | 'popular'
    sortOrder?: 'asc' | 'desc'
    minPrice?: number
    maxPrice?: number
    sizes?: string[]
    colors?: string[]
    inStock?: boolean
  } = {}) {
    const { db } = await connectToDatabase()

    let query: any = { isActive: true }

    if (options.category) {
      query.category = options.category
    }

    if (options.minPrice !== undefined || options.maxPrice !== undefined) {
      query.price = {}
      if (options.minPrice !== undefined) query.price.$gte = options.minPrice
      if (options.maxPrice !== undefined) query.price.$lte = options.maxPrice
    }

    if (options.sizes && options.sizes.length > 0) {
      query['sizes.name'] = { $in: options.sizes }
    }

    if (options.colors && options.colors.length > 0) {
      query['colors.name'] = { $in: options.colors }
    }

    if (options.inStock !== undefined) {
      query.inStock = options.inStock
    }

    let sort: any = {}
    const sortOrder = options.sortOrder === 'desc' ? -1 : 1

    switch (options.sortBy) {
      case 'price':
        sort.price = sortOrder
        break
      case 'rating':
        sort.rating = sortOrder
        break
      case 'newest':
        sort.createdAt = -1
        break
      case 'popular':
        sort.rating = -1
        break
      default:
        sort.createdAt = -1
    }

    const products = await db
      .collection('products')
      .find(query)
      .sort(sort)
      .skip(options.skip || 0)
      .limit(options.limit || 20)
      .toArray()

    const total = await db.collection('products').countDocuments(query)

    return {
      products: products as Product[],
      total,
      hasMore: (options.skip || 0) + (options.limit || 20) < total
    }
  }

  static async getProductById(id: string) {
    const { db } = await connectToDatabase()
    const product = await db.collection('products').findOne({
      _id: new ObjectId(id),
      isActive: true
    })
    return product as Product | null
  }

  static async getProductsByIds(ids: string[]) {
    const { db } = await connectToDatabase()
    const objectIds = ids.map(id => new ObjectId(id))
    const products = await db
      .collection('products')
      .find({
        _id: { $in: objectIds },
        isActive: true
      })
      .toArray()
    return products as Product[]
  }

  static async getCategories() {
    const { db } = await connectToDatabase()
    const categories = await db
      .collection('categories')
      .find({ isActive: true })
      .sort({ sortOrder: 1 })
      .toArray()
    return categories
  }

  static async searchProducts(query: string, options: {
    limit?: number
    category?: string
  } = {}) {
    const { db } = await connectToDatabase()

    const searchQuery: any = {
      isActive: true,
      $or: [
        { name: { $regex: query, $options: 'i' } },
        { description: { $regex: query, $options: 'i' } },
        { tags: { $in: [new RegExp(query, 'i')] } },
        { category: { $regex: query, $options: 'i' } }
      ]
    }

    if (options.category) {
      searchQuery.category = options.category
    }

    const products = await db
      .collection('products')
      .find(searchQuery)
      .limit(options.limit || 20)
      .toArray()

    return products as Product[]
  }

  static async recordProductView(productId: string, userId?: string, sessionId?: string, source: string = 'website') {
    const { db } = await connectToDatabase()

    const view: Omit<ProductView, '_id'> = {
      productId: new ObjectId(productId),
      userId: userId ? new ObjectId(userId) : undefined,
      sessionId: sessionId || 'anonymous',
      viewedAt: new Date(),
      source: source as any
    }

    await db.collection('productViews').insertOne(view)
  }

  static async getRecommendedProducts(userId?: string, category?: string, limit: number = 6) {
    const { db } = await connectToDatabase()

    // Simple recommendation logic - products from same category with high rating
    let query: any = { isActive: true, rating: { $gte: 4.0 } }

    if (category) {
      query.category = category
    }

    const products = await db
      .collection('products')
      .find(query)
      .sort({ rating: -1, reviewCount: -1 })
      .limit(limit)
      .toArray()

    return products as Product[]
  }
}
